runtime/src/mm/cpp/ObjectFactoryTest.cpp

```diff
@@ -317,6 +329,173 @@ TEST(ObjectFactoryStorageTest, EraseTheOnlyElement) {
+
+    EXPECT_THAT(actual, testing::ElementsAre(2, 3));
+    EXPECT_THAT(actualConsumer, testing::ElementsAre(1));
+}
+
+TEST(ObjectFactoryStorageTest, MoveMiddle) {
+    ObjectFactoryStorageRegular storage;
+    Producer<ObjectFactoryStorageRegular> producer(storage, SimpleAllocator());
+    Consumer<ObjectFactoryStorageRegular> consumer;
+
+    producer.Insert<int>(1);
+    producer.Insert<int>(2);
+    producer.Insert<int>(3);
+
+    producer.Publish();
+
+    {
+        auto iter = storage.Iter();
+        for (auto it = iter.begin(); it != iter.end();) {
+            if (it->Data<int>() == 2) {
+                iter.MoveAndAdvance(consumer, it);
+            } else {
+                ++it;
+            }
+        }
+    }
+
+    auto actual = Collect<int>(storage);
+    auto actualConsumer = Collect<int, alignof(void*)>(consumer);
+
+    EXPECT_THAT(actual, testing::ElementsAre(1, 3));
+    EXPECT_THAT(actualConsumer, testing::ElementsAre(2));
+}
+
+TEST(ObjectFactoryStorageTest, MoveLast) {
+    ObjectFactoryStorageRegular storage;
+    Producer<ObjectFactoryStorageRegular> producer(storage, SimpleAllocator());
+    Consumer<ObjectFactoryStorageRegular> consumer;
+
+    producer.Insert<int>(1);
+    producer.Insert<int>(2);
+    producer.Insert<int>(3);
+
+    producer.Publish();
+
+    {
+        auto iter = storage.Iter();
+        for (auto it = iter.begin(); it != iter.end();) {
+            if (it->Data<int>() == 3) {
+                iter.MoveAndAdvance(consumer, it);
+            } else {
+                ++it;
+            }
+        }
+    }
+
+    auto actual = Collect<int>(storage);
+    auto actualConsumer = Collect<int, alignof(void*)>(consumer);
+
+    EXPECT_THAT(actual, testing::ElementsAre(1, 2));
+    EXPECT_THAT(actualConsumer, testing::ElementsAre(3));
+}
+
+TEST(ObjectFactoryStorageTest, MoveAll) {
+    ObjectFactoryStorageRegular storage;
+    Producer<ObjectFactoryStorageRegular> producer(storage, SimpleAllocator());
+    Consumer<ObjectFactoryStorageRegular> consumer;
+
+    producer.Insert<int>(1);
+    producer.Insert<int>(2);
+    producer.Insert<int>(3);
+
+    producer.Publish();
+
+    {
+        auto iter = storage.Iter();
+        for (auto it = iter.begin(); it != iter.end();) {
+            iter.MoveAndAdvance(consumer, it);
+        }
+    }
+
+    auto actual = Collect<int>(storage);
+    auto actualConsumer = Collect<int, alignof(void*)>(consumer);
+
+    EXPECT_THAT(actual, testing::IsEmpty());
+    EXPECT_THAT(actualConsumer, testing::ElementsAre(1, 2, 3));
+}
+
+TEST(ObjectFactoryStorageTest, MoveTheOnlyElement) {
+    ObjectFactoryStorageRegular storage;
+    Producer<ObjectFactoryStorageRegular> producer(storage, SimpleAllocator());
+    Consumer<ObjectFactoryStorageRegular> consumer;
+
+    producer.Insert<int>(1);
+
+    producer.Publish();
+
+    {
+        auto iter = storage.Iter();
+        auto it = iter.begin();
+        iter.MoveAndAdvance(consumer, it);
```

Comment:
Checking that `it == iter.end()` here makes sense.

Reply:
Good point, also fixed the related test with `EraseAndAdvance`.